### ISSUE TITLE
validation layers fail to catch erroneous VkRenderPassMultiviewCreateInfo.pViewMasks

### ISSUE THREAD

**Issue by Phildo** [KhronosGroup/Vulkan-ValidationLayers#2319]
My bug was the following:

```
  VkRenderPassMultiviewCreateInfo renderPassMultiviewCreateInfo{};
  VkRenderPassMultiviewCreateInfo *prenderPassMultiviewCreateInfo{};
  if(vr)
  {
    uint32_t viewMask = 0b11;
    renderPassMultiviewCreateInfo.sType = VK_STRUCTURE_TYPE_RENDER_PASS_MULTIVIEW_CREATE_INFO;
    renderPassMultiviewCreateInfo.pViewMasks = &viewMask;
    renderPassMultiviewCreateInfo.subpassCount = 1;
    prenderPassMultiviewCreateInfo = &renderPassMultiviewCreateInfo;
  }
  VkRenderPassCreateInfo renderPassCreateInfo = {};
  renderPassCreateInfo.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
  renderPassCreateInfo.attachmentCount = (presentInfo->msaaSamples != VK_SAMPLE_COUNT_1_BIT) ? 3 : 2;
  renderPassCreateInfo.pAttachments = attachments;
  renderPassCreateInfo.subpassCount = 1;
  renderPassCreateInfo.pSubpasses = &subpassDescription;
  renderPassCreateInfo.pNext = prenderPassMultiviewCreateInfo;
  renderPassCreateInfo.dependencyCount = 1;
  renderPassCreateInfo.pDependencies = &subpassDependency
```

that is, `viewMask` was out of scope at the time its address was referenced in the renderPassCreateInfo. This caused my left eye to render black, and my right eye to flash black/white for a handful of frames before losing the device.

If the value at the address given by `pViewMasks` can be something which might eventually cause the device to be lost, I think that makes it a good candidate for validation layer coverage!

[Note: this took me a week to debug and I don't want anyone else to have to go through that lol]

**Comment by mark-lunarg** [comment]
@Phildo, we wish there was a practical way to check pointer validity.  There are code analysis tools such as valgrind or Visual Studio Code Analysis that can help to identify such situations, but it's well beyond the capabilities of this tool.

**Comment by Phildo** [comment]
@mark-lunarg I agree that "pointer validity" is reasonably out of scope for a validation layer. however, if the pointer is pointing to _invalid data which might result in a crash_, I believe that is _within scope_ to test.

to be clear, there are two possibilities here:

A. my bug arose from a UB pointer dereferencing, and the dereferencing itself set off a chain of events which eventually resulted in a crash. this would be out of scope for the validation layers.

B. my bug arose _not_ from the pointer dereferencing itself, but from the fact that the garbage data at the other end of the dereference communicated an invalid viewmask. I _believe_ this might be within validation layer scope?

I _assume_ B was my problem, though am not 100% sure. Regardless, I think B is a possibility that could be captured by the validation layers.

**Comment by mark-lunarg** [comment]
@Phildo, after some digging and conferencing, we're not positive that this could be done statically (or at all, usefully).  In any case, we came across nothing in the Vulkan spec nor in the SPIIR-V spec that says this is invalid behavior, so there's little we can do.  A possible next step might be to open an issue at [Vulkan-Docs](https://github.com/KhronosGroup/Vulkan-Docs).

**Comment by Phildo** [comment]
Ah, interesting. So it's technically a bug in the _runtime_ that it crashed, as the spec doesn't disqualify any uint32_t as a viewMask, regardless of other context (such as "views available").

If that's the case, I'm unsure what change is needed in the docs... Unless by "open an issue in the docs" you're implying I "open an issue for the _spec_ to _make_ it disqualify invalid uint32_ts"?

(sorry if I'm being pedantic- I'm new to big open source projects/documentation and don't want to waste any more of your or anyone else's time because I'm misunderstanding something!)

**Comment by baldurk** [comment]
What about `VUID-VkFramebufferCreateInfo-renderPass-04536`?

> If `renderPass` was specified with non-zero view masks, each element of `pAttachments` that is used as an input, color, resolve, or depth/stencil attachment by `renderPass` **must** have a `layerCount` greater than the index of the most significant bit set in any of those view masks

Which could verify at framebuffer create time that the view mask which was read (unknowingly from an invalid pointer) has no invalid bits set. Curiously the imageless framebuffer version `VUID-VkFramebufferCreateInfo-renderPass-03198` is checked, but this one isn't.

My guess is that the device lost is because the viewMask actually had bits that were set which were out of bounds for the attachments array size. If the view mask ended up with only valid bits set (or it was 0) then while the behaviour might not have been what the code originally intended I wouldn't necessarily expect a crash either.

**Comment by mark-lunarg** [comment]
@Phildo, no problem at all, this is a complicated issue at best.  I'm don't have a working knowledge of the MultiView extension, but as I understand it the if the viewmask matches the ViewIndex decoration in the shader, then it Does Something to those layers of the framebuffer attachment.  It might be that there should be a valid usage statement that says something like 'no index set in pViewMask must be greater than the number of layers in the target framebuffer attachment'.  I have no idea if this makes any sense, but that would the the sort of thing you could ask in Vulkan-Docs.

**Comment by mark-lunarg** [comment]
@baldurk, thanks!  I was searching for `pViewMasks `and totally missed this.  @Phildo, sorry for the confusion.

**Comment by mark-lunarg** [comment]
Leaving a note here for the eventual implementer, closed PR #1142 contains checks for another handful of unimplemented and related VUIDs.

**Comment by ncesario-lunarg** [comment]
This is now tracked under https://github.com/KhronosGroup/Vulkan-ValidationLayers/projects/4#card-68681483.
